Feat/worldpay

Dyl4n-C0de opened the pull request:
Error Handling, Auth, Max Transactions, Duplicate Checks
- Implemented error handling for duplicate merchant email and phone number.
- Set default value to an empty string in `index.ts` for unspecified fields.
- Added explanatory comments regarding JWT and authentication methods.
- Enhanced error handling in `src/v1/services/Transaction.ts`:
  - Added a check at lines 48-49 to throw an error if the BIN is blacklisted.
  ~~- Enclosed the `switch (processor)` logic within a try block to manage errors when the transaction amount exceeds `maxTransactionAmount`.~~
- Defined the maximum transaction amount in the merchant schema to enforce limits.

thomas779 commented:
@Dyl4n-C0de Let's avoid doing repo forking over branches. There is a risk of exposing creds: the repo in the org is private; on your profile it's public. I've only ever seen repo forking used in OSS development, rarely within product teams. Even then, it is usually for big, refactor changes over adding middleware feats or fixes.

thomas779 commented on pnpm-lock.yaml:
this file should be deleted

Dyl4n-C0de replied:
added it to .gitignore in the fix earlier, but need to remove it entirely. good catch.

thomas779 commented on src/config/index.ts:
```diff
@@ -42,11 +43,11 @@ export const Config = {
-      username: loadEnvironmentVariable('WORLDPAY_USERNAME'),
-      password: loadEnvironmentVariable('WORLDPAY_PASSWORD'),
-      accessCheckoutId: loadEnvironmentVariable('WORLDPAY_ACCESS_CHECKOUT_ID'),
+      apiUrl: loadEnvironmentVariable('WORLDPAY_API_URI', ''),
+      entityRef: loadEnvironmentVariable('WORLDPAY_ENTITY_REF', ''),
+      username: loadEnvironmentVariable('WORLDPAY_USERNAME', ''),
+      password: loadEnvironmentVariable('WORLDPAY_PASSWORD', ''),
+      accessCheckoutId: loadEnvironmentVariable('WORLDPAY_ACCESS_CHECKOUT_ID', ''),
```
The defaults should not be empty. 
```suggestion
      apiUrl: loadEnvironmentVariable('WORLDPAY_API_URI'),
      entityRef: loadEnvironmentVariable('WORLDPAY_ENTITY_REF'),
      username: loadEnvironmentVariable('WORLDPAY_USERNAME'),
      password: loadEnvironmentVariable('WORLDPAY_PASSWORD'),
      accessCheckoutId: loadEnvironmentVariable('WORLDPAY_ACCESS_CHECKOUT_ID'),
```

thomas779 commented on src/v1/handlers/merchant.ts:
```diff
@@ -18,6 +18,11 @@ export async function createMerchantHandler(
   rep: FastifyReplyTypebox<typeof MerchantCreateSchema>
 ): Promise<void> {
   try {
+    const emailExists = await merchantService.emailExists(req.body.email);
+    if (emailExists) {
+      return errorResponse(req, rep, ERROR400.statusCode, 'A merchant with this email already exists.');
+    }
+
```
this should be in the middleware

thomas779 commented on src/v1/schemas/merchant.ts:
```diff
@@ -19,6 +19,7 @@ export const MerchantCreateSchema = {
       state: t.Optional(t.String({ minLength: 2, maxLength: 2 })),
       country: t.String({ minLength: 2, maxLength: 2 }),
     }),
+    maxTransactionAmount: t.Number({ maximum: 10000 }), // Maximum transaction amount in dollars
```
maximum amount should not be in the schema

thomas779 commented on src/v1/services/Merchant.ts:
```diff
@@ -24,6 +24,15 @@ export class MerchantService {
   }
 
   /** @dev create partner */
+  public async emailExists(email: string): Promise<boolean> {
+    const count = await prisma.merchant.count({
+      where: {
+        email: email,
+      },
+    });
+    return count > 0;
+  }
+
```
I like this!

thomas779 commented on src/v1/services/Merchant.ts:
```diff
@@ -41,6 +50,18 @@ export class MerchantService {
       registeredAddress;
 
     try {
+      const existingEmailCount = await prisma.merchant.count({ where: { email } });
+      if (existingEmailCount > 0) {
+        console.error(`Email already exists in the database: ${email}`);
+        throw new Error('A merchant with this email already exists.');
+      }
+
+      const existingPhoneCount = await prisma.merchant.count({ where: { phoneNumber } });
+      if (existingPhoneCount > 0) {
+        console.error(`Phone number already exists in the database: ${phoneNumber}`);
+        throw new Error('A merchant with this phone number already exists.');
+      }
+
```
No need to check if the email exists twice. Just write existingEmailCount() and existingPhoneCount() as separate functions that are called in the middleware.

thomas779 commented on src/v1/services/Merchant.ts:
```diff
@@ -24,6 +24,15 @@ export class MerchantService {
 
   /** @dev create partner */
+  public async emailExists(email: string): Promise<boolean> {
+    const count = await prisma.merchant.count({
```
What were the tradeoffs you made between `findUnique` vs. `findFirst` vs. `count`?

thomas779 reviewed (changes requested):
Good first PR. Made a couple of comments to your code. Would recommend you delete your repo fork and make a new branch; copy the main/staging branch instead of the feat/worldpay branch.

Dyl4n-C0de commented:
@thomas779 Excellent. I implemented changes based on the review, as well as, added more detailed regex validation and walletAddress checks in the new branch "feat/merchant-transactions"
